Notes for Exercise-12 (#4)
mike-north/ember-octane-workshop: notes/12-guarding-routes.md

# Guarding Routes

In this exercise, we will be adding redirection in the UI flow based on the availability of an authenticated session.
Specifically, users should be able to access the `login` route, only when an authenticated session _does not exist_. And similarly, users should be able to access the `teams` route only when an authenticated _session exists_.

So let's get started.

> ## Changes Needed in Javascript

Reviewer: I think we should (1) make this title more descriptive and (2) break the exercise up into discrete "tasks" 

- "Enhancing the `auth` service"
- "Adding a redirect to the `/teams` page"
- "Adding a redirect to the `/login` page"

> First let's implement the auth service that we will be using to do session validation(or invalidation). This auth service is defined at `app/services/auth.js`.

Reviewer: They've already been working with an auth service since step 10/11. Maybe change this to
```suggestion
First let's enhance the auth service so that we can easily tell if there's a logged in user, and implement a `logout` action. This auth service is defined at [`app/services/auth.js`](../app/services/auth.js). 
```

Import the action decorator at the top of the file:

```js
import { action } from '@ember/object';
```

> Then add a property, `isAuthenticated` that will be used to store if the session is in an authenticated state, and an action method, `logout`, for invalidating the authenticated session and signing off the user.

Reviewer: ```suggestion
Then add a getter-based property, `isAuthenticated` that will be `true` if the user is logged in and `false` otherwise. Add an action called `logout` that clears the user ID we write to localStorage on login, and sends the user to the `/login` page.
```
Simplifying some of the language. I'm hopeful we can avoid talking about a "session" and just talk in terms of an ID stored somewhere.

```js
  get isAuthenticated() {
    return !!this.currentUserId;
  }

  @action
  logout() {
    window.localStorage.removeItem(AUTH_KEY);
    this.router.transitionTo('login');
  }
```

> In the login route, present at `app/routes/login.js`, add a `beforeModel` hook to perform session validations and handle the UI flow accordingly.

Reviewer: ```suggestion
In the login route, present at [`app/routes/login.js`](../app/routes/login.js), add a `beforeModel` hook to check whether the user is authenticated, and either redirect the user or not.
```

> In the login route, inject the imported `auth` service and add the `beforeModel` hook to perform session validation.

Reviewer: ```suggestion
In the login route, inject the imported `auth` service and add the following `beforeModel` hook.
```
add link to ember docs for `beforeModel`

```js
   /**
   * @type {AuthService}
   */
  @service auth;

  async beforeModel(transition) {
    await super.beforeModel(transition);
    if (this.auth.isAuthenticated) {
      this.transitionTo('teams');
    }
  }
```

> And, in the same route file(`login` route), import the `auth` service, since its not already available(injected).

Reviewer: Let's instruct readers to import first, then implement (so they get the nice autocomplete as they type).

We should also point out that `AuthService` is used in the JSDoc comment, to improve the editing experience (i.e., `this.auth.*` has a nice autocomplete).

```js
import { inject as service } from '@ember/service';
import AuthService from 'shlack/services/auth';
```

Now let's move on to `teams` route.

> Similar to `login` route, handle session validation in the `beforeModel` hook of the `teams` route after injecting the auth service into the route. There is a key difference to note here - the validation logic is flipped, that is the redirection will happen only when an authenticated session _does not exists_

Reviewer: ```suggestion
The `login` route should have a similar `beforeModel` hook, but note that the validation logic is flipped. We redirect to the `/login` page only the user is _logged out_.
```

In the `teams` route, inject the imported `auth` service and add the `beforeModel` hook to perform session validation.

```js

> // app/routes/teams.js

Reviewer: Let's make these clickable links instead of comments in the code

```md
[`app/routes/teams.js`](../app/routes/teams.js)

```

   /**
   * @type {AuthService}
   */
  @service auth;

  async beforeModel(transition) {
    await super.beforeModel(transition);
    if (!this.auth.isAuthenticated) {
      this.transitionTo('login');
    }
  }
```

> And, in the same route file(`teams` route), import the `auth` service, since its not already available(injected).

Reviewer: again, deal with imports before methods

```js
// app/routes/teams.js
import { inject as service } from '@ember/service';
import AuthService from 'shlack/services/auth';
```

## Changes in Templates

Now that we have the javascript implementation in place, lets add the UI for the same.
In `app/templates/components/team-sidebar.hbs`, replace the `LinkTo` component with a plain old HTML `button` element with an onclick handler that will trigger the `logout` action that you defined in `app/services/auth.js`.

```diff
-    <LinkTo @route='login' {{! destination route }}
-            @tagName="button" {{! use <button> instead of <a> }}
-            class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button" {{! HTML classes}}

> +    <button {{on "click" this.auth.logout}} class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button"

Reviewer: I think I'll adjust this solution so the `class=` is on its own line. This will make your diff a bit simpler, since that same line is present in both the `<button>` and `<LinkTo>`. Let's update this diff accordingly 

```diff
-    <LinkTo @route='login' {{! destination route }}
-            @tagName="button" {{! use <button> instead of <a> }}
-			 {{! HTML classes}}
             class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button"
+    <button {{on "click" this.auth.logout}}
			 class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button"
     >
```

Author: How do you recommend we move forward with this one ? Should I wait till code change is made to exercise solution ?

+    >
      Logout
-    </LinkTo>
+    </button>
```

> ## Adding Tests

Reviewer: Can we break this up into a few discrete tasks? Maybe one task for the `login-test` and one for `logout-test`?

> Now that we have the implementation in place for this exercise, lets add some tests for the same.

Reviewer: ```suggestion
Now that we have the implementation in place for our redirect logic, lets add some tests for the same.
```

> First, let's add some acceptance tests to make sure the UI interactions are working as expected.

Reviewer: Maybe comment on why acceptance tests are an appropriate choice for this change to the application?

> In `tests/acceptance/login-test.js`, import the `StubbedAuthService` service:

Reviewer: filename -> clickable link

```js
import StubbedAuthService from '../test-helpers/auth-service';
```

Then, inject the auth service inside the `beforeModel` hooks for test setup.

```js
hooks.beforeEach(function() {
  this.owner.register('service:auth', StubbedAuthService);
});
```

Modify the test with label, `starting logged out` as follows:

```js
test('starting logged out, then logging in', async function(assert) {
  const auth = this.owner.lookup('service:auth');
  auth.currentUserId = null;

  await visit('/login');
  assert.equal(currentURL(), '/login');

  await fillIn('select', '1');
  await click('form input[type="submit"]');

  assert.equal(currentURL(), '/teams');
});
```

Then add a test for the use case when the user is already logged in, as follows:

```js
test('already logged in', async function(assert) {
  const auth = this.owner.lookup('service:auth');
  auth.currentUserId = '1';

  await visit('/login');

  assert.equal(currentURL(), '/teams');
});
```

> Now let's add acceptance tests to test when users for logged out. The test file is present at `tests/acceptance/logout-test.js`. And as in the previous test, first import the `StubbedAuthService` service:

Reviewer: filename -> clickable link

```js
import StubbedAuthService from '../test-helpers/auth-service';
```

> Then, remove the definition for test with label, 'visiting /teams'.

Reviewer: ```suggestion
Remove the definition for test with label, 'visiting /teams'.
```
